RFCs/0020-User-Specified-TestAdapter-Lookup.md

```diff
@@ -0,0 +1,41 @@
+# 0020 User Specified Test Adapter Lookup
+
+## Summary
+This note outlines the proposed changes for lookup and initialization of the test adapters, based on User input.
+
+## Motivation
+
+As per the current adapter lookup model, when ever a Run/Discovery call is made from IDE to TestPlatform, the test adapter initialization happens in multiple phases listed below
+
+* InitializeExtensions(): In this call, the IDE passes list of all adapters it finds in current session, these include adapters referenced via nuget in entire solution, & adpaters acquired via Vsix(e.g. GoogleTestAdapter, BoostAdapters, etc..)
+* TestAdapterPath: After InitializeExtensions call, TestPlatform then looks into adapters present in TestAdapterPath mentioned in RunSettings, & adds adapters to existing list.
+* TestPlatform then adds default Adapters from **Extensions** directory to current list.
+* Finally platform looks in adapters present in test source directory, & adds them to adapter list.
+
+Once adapters are gathered from all the possible locations, this list is passed to testhost process, which then loads all these adapters, & passes test source(s) to all the adapters.
+
+Following the above adapter look up logic, we pass down test sources to adapters which have no business running them, for e.g. passing a managed test source to a native adapter(GoogleTestAdapter, BoostAdapters, etc.), & vice-versa.
+
+## Principles
+1. The adapter referenced in the project should be used for test execution
+2. Performance should not degrade.
+3. Adapter lookup logic is consistent across runs.
+4. Adapter lookup logic should be clear and consistent for both IDE/Editor and CLI runs.
+
+## Proposed Changes
+Here are the proposed changes based on the sources available for lookup of the adapters.
+User can restrict adapter lookup to only adapters close to source location, & those present in TestAdapterPath under RunSettings, via setting **UseSpecifedAdapterLocations** flag in RunSettings to **true**.
```

Comment:
nit: UseSpecifiedAdapterLocations

I find above a little confusing, as it does not tell me that adapters in the test source directory will be used.
Should we call it UseLocalAdapters ?

How about we go for a SkipDefaultAdapters instead ?

Reply:
It's a valid nit, IMO. The user has specified what adapters to use in TestAdapterPath and/or RunSettings. They will then wonder "Why do I have to say to use what is specified?" However "UseLocalAdapters is also confusing. Perhaps "UseOnlyLocalAdapters" or "ExcludeDefault(?)Adapters".

Reply:
+1 to the nit. For flag name we should use "UseUserSpecifiedAdapters" and in the VS IDE UI we should use text explaining this option - "Use the test adapters directly referenced by test projects or ones specified in the RunSettings file"

•	Yes we should have an option ( a check box)  in the UI to enable this flag. This should be a per solution setting and only for C# and VB as other languages are shipping and plan to ship adapters as vsix in the box ( like C++ and TS). This box should be checked by default.
•	From and IDE perspective, we should not spawn a new host for each test assembly. 
•	If no adapters are found when the above option is selected then we should probably not fall back. Instead we should educate users to uncheck this box. This box should be checked by default though.